From 5cd75bea7e7712425afabfe311cfb149cb9ddc1a Mon Sep 17 00:00:00 2001
From: anonym <anonym@riseup.net>
Date: Mon, 8 Feb 2021 15:07:29 +0100
Subject: [PATCH] Enable the Tor Seccomp sandbox if possible.

I.e. as long as we are not using pluggable transports, enable the
sanbox. These changes to Tor Launcher only call the tor-sandbox-helper
in the appropriate way, but it is that script that actually
enables/disables the sandbox.

Refs: #17330
---
 src/chrome/content/network-settings.js | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/src/chrome/content/network-settings.js b/src/chrome/content/network-settings.js
index 81d7138..4dafb26 100644
--- a/src/chrome/content/network-settings.js
+++ b/src/chrome/content/network-settings.js
@@ -600,6 +600,20 @@ function onWizardFirstPanelConnect()
     removeSettingsAndConnect()
 }
 
+function torSandboxSetter(enable)
+{
+  var status = enable ? "1" : "0";
+
+  var helper = Cc['@mozilla.org/file/local;1'].createInstance(Ci.nsIFile);
+  helper.initWithPath('/usr/bin/sudo');
+
+  var process = Components.classes["@mozilla.org/process/util;1"]
+                .createInstance(Components.interfaces.nsIProcess);
+  process.init(helper);
+
+  var args = ['/usr/local/lib/tor-sandbox-helper', status];
+  process.run(true, args, args.length);
+}
 
 function removeSettingsAndConnect()
 {
@@ -2325,6 +2339,12 @@ function initDefaultBridgeTypeMenu()
   }
 }
 
+// Returns true iff bridge is a normal bridge, not using any pluggable
+// transport.
+function isNormalBridge(bridge)
+{
+  return bridge.replace(/^bridge\s+/i, "").split(/\s+/)[0].match(/^[0-9.:]+$/) != null;
+}
 
 // Returns true if settings were successfully applied.
 function applyBridgeSettings(aUseDefaults)
@@ -2337,6 +2357,11 @@ function applyBridgeSettings(aUseDefaults)
   if (aUseDefaults)
     TorLauncherUtil.setCharPref(kPrefDefaultBridgeType, "");
 
+  let useBridges = settings[kTorConfKeyUseBridges];
+  let bridgeList = settings[kTorConfKeyBridgeList];
+  let onlyNormalBridges = bridgeList ? bridgeList.every(isNormalBridge) : true;
+  let enableSandbox = !useBridges || onlyNormalBridges;
+  torSandboxSetter(enableSandbox);
   return setConfAndReportErrors(settings, "configureSettings");
 }
 
-- 
2.30.0

